### PR TITLE
Add detection of confidential virtualization on s390x architcture

### PR DESCRIPTION
**Pull request by berrange**
This small series primarily adds detection of s390x protected virtualization (aka secure execution), but also fixes some docs issues.

### CODE REVIEW

**Comment by berrange** [comment]
NB, I tested the output of 'systemd-detect-virt --cvm' on an s390x RHEL 9.5 guest with protvirt enabled.

File: src/basic/confidential-virt.c [hunk 1]
```diff
@@ -209,6 +209,23 @@ static ConfidentialVirtualization detect_confidential_virtualization_impl(void)
 
         return CONFIDENTIAL_VIRTUALIZATION_NONE;
 }
+#elif defined(__s390x__)
+static ConfidentialVirtualization detect_confidential_virtualization_impl(void) {
+        int fd;
+        char c;
+
+        fd = open("/sys/firmware/uv/prot_virt_guest", O_RDONLY);
```
Comment (bluca) [inline src/basic/confidential-virt.c]: please use the `read_full_virtual_file()` helper

**Review by bluca — APPROVED** [review]

File: src/basic/confidential-virt.c (hunk 1 above)
Comment (berrange) [reply src/basic/confidential-virt.c]: Done

File: src/basic/confidential-virt.c [hunk 2]
```diff
@@ -194,40 +195,62 @@ static bool detect_hypervisor(void) {
...
+}
+#elif defined(__s390x__)
+static ConfidentialVirtualization detect_confidential_virtualization_impl(void) {
+        _cleanup_free_ char *s = NULL;
+        size_t readsize;
+        int r;
+
+        r = read_full_virtual_file("/sys/firmware/uv/prot_virt_guest", &s, &readsize);
+        if (r < 0) {
+                log_debug_errno(r, "Unable to read /sys/firmware/uv/prot_virt_guest: %m");
```
Comment (bluca) [inline src/basic/confidential-virt.c:221]: I imagine this file is only present on certain s390x setups? If so, the log should be behind an `if (r != -ENOENT)` as there's no point in logging about it not being found

Comment (berrange) [reply src/basic/confidential-virt.c:221]: The kernel made it a bit wierd in that the entire `/sys/firmware/uv` tree will be absent if no ultravisor is present, but when the ultravisor is present, this subtree will be present for both host and guest kernels.

* Old kernel -> file absent
* New kernel + no ultravisor -> subtree absent in both host and guest
* New kernel + ultravisor -> subtree present in both host and guest. 'prot_virt_host' and 'prot_virt_guest' files distinguish host vs guest context.

IOW, i think having the debug log even for ENOENT is desirable, as that indicates that we've checked for the file and found it absent, confirming a non-confidential guest.

Comment (bluca) [reply src/basic/confidential-virt.c:221]: ok

File: man/systemd-detect-virt.xml [hunk 3]
```diff
@@ -62,7 +62,7 @@
         </thead>
         <tbody>
           <row>
-            <entry valign="top" morerows="16">VM</entry>
+            <entry valign="top" morerows="17">VM</entry>
```
Comment (yuwata) [inline man/systemd-detect-virt.xml:65]: Ouch, thanks.

**Review by yuwata — APPROVED** [review]
LGTM.

**Comment by yuwata** [comment]
Please backport the third commit.

**Comment by berrange** [comment]
> Please backport the third commit to stable.

@yuwata submitted https://github.com/systemd/systemd/pull/33945

**Pull request merged**